feat: add GitHub Actions workflows for CI/CD

kdcokenny opened the pull request:
## Summary
- Add comprehensive GitHub Actions workflows for CI/CD pipeline
- Implement testing, linting, and PyPI publishing automation
- Use modern best practices with uv package manager and trusted publishing

## Workflows Added

### 1. CI Workflow (`ci.yml`)
- Runs on every push to main and pull requests
- Matrix testing across Python 3.10, 3.11, and 3.12
- Full test suite execution
- Linting with ruff and type checking with mypy
- Package building and verification
- Optimized caching with uv

### 2. Code Quality Workflow (`code-quality.yml`)
- Dedicated PR checks for code quality
- Automated formatting and linting verification
- Comments on PRs with specific issues and fixes
- Helps maintain consistent code style

### 3. Publish Workflow (`publish.yml`)
- Automated PyPI publishing on GitHub releases
- Uses PyPI's trusted publishing (no API tokens needed)
- Optional TestPyPI publishing for testing
- Sigstore signing for package attestations
- Uploads signed packages to GitHub releases

## Changes Made
- Fixed test URLs from `edge-alpha.postcrawl.com` to `edge.postcrawl.com`
- Removed coverage requirements to simplify test runs
- Fixed package verification to use virtual environments

## ✅ All Checks Passing
The workflows are now working correctly as demonstrated by this PR.

## Next Steps Required

### 1. Configure PyPI Trusted Publishing
1. Go to https://pypi.org/manage/project/postcrawl/settings/publishing/
2. Add a new trusted publisher with these details:
   - **Owner**: `post-crawl`
   - **Repository name**: `python-sdk`
   - **Workflow name**: `publish.yml`
   - **Environment name**: `pypi`

### 2. Set Up GitHub Environments
1. Go to Settings → Environments in this repository
2. Create `pypi` environment
   - Consider adding protection rules (e.g., required reviewers)
3. Create `testpypi` environment for testing

### 3. Optional Enhancements
- **Branch Protection**: Consider requiring CI checks to pass before merging

## Benefits
- ⚡ Fast CI/CD with uv package manager
- 🔒 Secure publishing without storing API tokens
- 🧪 Comprehensive testing across Python versions
- 📝 Automated code quality enforcement
- 📦 Signed packages with attestations

kdcokenny committed:
feat: add GitHub Actions workflows for CI/CD

- Add CI workflow for testing, linting, and building
- Add code quality workflow for PR checks
- Add publish workflow for PyPI releases with trusted publishing
- Test across Python 3.10, 3.11, and 3.12
- Use uv package manager for fast builds
- Implement caching optimization
- Include Sigstore signing for releases

kdcokenny committed:
fix: remove coverage requirements and fix package verification

kdcokenny committed:
fix: correct API URLs in tests from edge-alpha to edge

kdcokenny committed:
Merge branch 'main' into feat/github-actions-ci-cd

kdcokenny committed:
feat: Update GitHub Actions to use uv publish directly

- Replace pypa/gh-action-pypi-publish with native uv publish commands
- Update package verification in CI to match Makefile approach
- Maintain OIDC trusted publishing for secure, token-free deployments
- Explicitly pass GitHub OIDC environment variables to uv publish

This aligns our CI/CD pipeline with the project's use of uv throughout
the development workflow and follows uv's best practices for publishing.

kdcokenny commented:
## Updated GitHub Actions to Use uv publish

I've updated the workflows to use `uv publish` directly instead of the PyPA publishing action, following uv's best practices.

### Changes Made

#### 1. **publish.yml Updates**
- **TestPyPI Publishing**: Now uses `uv publish --index testpypi`
- **PyPI Publishing**: Now uses `uv publish` 
- Both jobs now checkout the repository and install uv before publishing
- Environment variables for OIDC tokens are explicitly passed to enable trusted publishing

#### 2. **ci.yml Updates**
- **Package Verification**: Updated to use the same approach as the Makefile:
  ```bash
  uv run --with postcrawl --no-project --refresh-package postcrawl -- python -c "import postcrawl"
  ```

### Trusted Publishing Setup

The workflows continue to use GitHub's OIDC (OpenID Connect) for trusted publishing:
- No API tokens need to be stored as secrets
- The `id-token: write` permission enables OIDC token generation
- The environment variables are automatically provided by GitHub

### Prerequisites

Before these workflows will work, you need to configure trusted publishing:

1. **For TestPyPI**:
   - Go to https://test.pypi.org/manage/project/postcrawl/settings/publishing/
   - Add a new trusted publisher with:
     - Owner: `post-crawl`
     - Repository: `python-sdk`
     - Workflow: `.github/workflows/publish.yml`
     - Environment: `testpypi`

2. **For PyPI**:
   - Go to https://pypi.org/manage/project/postcrawl/settings/publishing/
   - Add a new trusted publisher with:
     - Owner: `post-crawl`
     - Repository: `python-sdk`
     - Workflow: `.github/workflows/publish.yml`
     - Environment: `pypi`

### Benefits

1. **Consistency**: Uses uv throughout the entire workflow
2. **Security**: No need to store API tokens as secrets
3. **Simplicity**: Direct use of uv commands matches local development
4. **Reliability**: uv handles retries and duplicate uploads gracefully

kdcokenny committed:
fix: Mark TestPyPI index as explicit to prevent CI conflicts

The TestPyPI index has outdated packages that conflict with our build
dependencies. By marking it as explicit, it will only be used when
specifically requested with --index testpypi, not during regular
dependency resolution in CI.

ps428 commented on .github/workflows/publish.yml:
```diff
@@ -0,0 +1,144 @@
+      - name: Publish to PyPI
+        env:
+          # These are provided by GitHub for OIDC trusted publishing
+          ACTIONS_ID_TOKEN_REQUEST_TOKEN: ${{ env.ACTIONS_ID_TOKEN_REQUEST_TOKEN }}
```
although environments are different for these two (testpypi & pypi), please validate the tokens are correctly mapped.

ps428 replied:
configured pypi and testpypi OIDC. please update the publish.yml based on the official docs: https://docs.github.com/en/actions/how-tos/security-for-github-actions/security-hardening-your-deployments/configuring-openid-connect-in-pypi

ps428 reviewed (commented):
rest lgtm

kdcokenny committed:
feat: update GitHub Actions to use latest uv version and best practices

- Update uv to v0.7.20 and setup-uv action to v6 across all workflows
- Simplify publish workflow by removing manual OIDC token configuration
- Enable built-in caching with setup-uv action
- Align workflow structure with official uv documentation
- Trust uv's internal OIDC handling for PyPI/TestPyPI publishing

kdcokenny commented:
@claude review

ps428 reviewed (approved):
Looks good.
